Updates edit_notes.py, removes global declarations, adds columns to csv

kemartin2015 opened the pull request:
I tried to address the deprecated ASnake endpoint issue, remove the global declarations, and add more information to the csv file to make it more consistent with the csv file output in the notes_csv.py script.

So far, I'm consistently getting error messages in line 25 where I'm trying to get json for notes each time I try a different approach. Could you offer some guidance on how to fix this line? I'm also unsure about my changes to lines 42 and 52.

helrond commented on archivessnake/edit_notes.py:
```diff
@@ -78,20 +76,18 @@ def get_parser():
-
 if __name__ == "__main__":
-    main()
+    client = ASpace().client
+    spreadsheet_path = os.path.join(os.path.abspath(os.path.dirname(__file__)), OUTPUT_FILENAME)
+    writer = csv.writer(open(spreadsheet_path, "w"))
+    main(client, writer)
```
I would suggest moving all of this up to the `main` method - there's no reason to do some instantiation here and it just forces you to have to pass those arguments up to main.

helrond commented on archivessnake/edit_notes.py:
```diff
@@ -7,61 +7,59 @@
-        if args.level in [record.level, "all"]:
-            notes = aojson.get("notes")
+        if args.level in [record['level'], "all"]:
+            notes = client.get(record['notes']).json()
```
I think the issue here is that `client` is not being passed from `main` so this function doesn't know how to execute `walk_tree`. You're _probably_ getting an error in line 25 because whatever you're getting back from `walk_tree` doesn't have a `json` method.

I would actually suggest passing in the `args` and the `client` to this method. I believe that you can pass a URI _or_ a full JSON object to `walk_tree` so you may be able to simplify a bit there too.

helrond commented on archivessnake/edit_notes.py:
```diff
@@ -7,61 +7,59 @@
 
 
-def save_record(uri, data):
+def save_record(record, data):
```
You're going to need to pass in `client` as well here.

helrond commented on archivessnake/edit_notes.py:
```diff
@@ -7,61 +7,59 @@
                             if args.action == "delete":
                                 del notes[idx]
                                 print("{} note deleted from object {}".format(
-                                    args.note_type, record.uri))
+                                    args.note_type, record['uri']))
                             if args.action == "modify":
                                 print("{} note was originally {} and was changed to {} in object {}".format(
-                                    args.note_type, content, args.replace_string, record.uri))
-                                subnote["content"] = args.replace_string
+                                    args.note_type, content, args.replace_string, record['uri']))
+                                subnote['content'] = args.replace_string
```
I would consider breaking this out into its own method, something like `handle_matching_note`?

helrond reviewed (commented):
I added some specific suggestions below but I'm also wondering if it would make sense to convert this to an object-oriented approach? It seems like there are a bunch of things you need to pass around that could be attributes on an object rather than args in functions.

helrond commented on archivessnake/edit_notes.py:
```diff
@@ -7,61 +7,59 @@
-spreadsheet_path = os.path.join(os.path.abspath(os.path.dirname(__file__)), OUTPUT_FILENAME)
 
-def process_tree(args, resource):
+def process_tree(args, resource_id):
```
I'm not sure it's a good idea to pass in `args` like this and then reference the specific attributes further down. I know it looks cleaner, but it masks the arguments that should be there, and to me makes the script a little harder to understand.

kemartin2015 commented:
Hi @helrond, thank you for these suggestions. I worked on addressing your comments, but I still don't think I'm passing client correctly because I'm getting this error pretty consistently:

`raise Exception('Could not resolve "{}" to a URI'.format(thingit))
Exception: Could not resolve "Namespace(note_type='accessrestrict', action='delete', resource_id=13291, search_string='restrict', level='all', replace_string=None)" to a URI`

I agree that it makes sense to change this script to an object-oriented approach. It's been a challenge to jump back in and make changes with all of the arguments that need to be passed around.

helrond commented on archivessnake/edit_notes.py:
```diff
@@ -82,16 +81,13 @@ def main():
-    create_spreadsheet(["Box Number", "Archival Object URI"])
-    process_tree(args, aspace.repositories(AS_REPOSITORY).resources(args.resource_id))
+    create_spreadsheet(writer, ["Collection Title", "Finding Aid Number", "URI", "Ref ID", "Object Title", "Box Number"])
+    process_tree(client.get(f'/repositories/2/resources/{(args.resource_id)}').json(), parser)
```
Here you are passing a JSON instance of a resource, but `process_tree` is expecting to be called with the `client` itself.

kemartin2015 commented:
Hi @helrond, I made quite a few changes, and the script is working now. Could you take another look whenever you get a chance?

helrond commented on archivessnake/edit_notes.py:
```diff
@@ -7,64 +7,67 @@
+def process_tree(client, parser, writer):
     """Iterates through a given collection, file, or series provided by user input. Finds note content that matches user input and then deletes or modifies relevant notes according to user preference.  """
-    for record in resource.tree.walk:
+    args = parser.parse_args()
```
I know this allows you to pass fewer arguments, but I think it's a bit of an anti-pattern in that there's no clear relationship between the arguments you're creating in `get_parser` and then what's being used throughout the rest of the script. I would prefer to see something slightly different:

- set and parse the args as the _first_ thing in the script. You can do this in the first few lines of `main` OR even in the block below (lines 96). That way, if someone doesn't pass the expected arguments, the script doesn't do anything. As it is, it will try and execute a bunch of stuff before actually parsing the arguments.
- Pass ALL the arguments you need into `process_tree`. This will mean you have a lot of arguments, but that's okay!
- Pass just the arguments you need into other methods (like `handle_matching_notes` etc)

That approach makes clear what's being used where, and also is much easier to write unit tests against.

helrond commented on archivessnake/edit_notes.py:
```diff
@@ -7,64 +7,67 @@
     """Iterates through a given collection, file, or series provided by user input. Finds note content that matches user input and then deletes or modifies relevant notes according to user preference.  """
-    for record in resource.tree.walk:
+    args = parser.parse_args()
+    resource_id = client.get(f'/repositories/2/resources/{(args.resource_id)}').json()
```
This variable is a bit misleading (especially since one of your args has the same name) because its really the full json of the resource, not the ID.

helrond commented on archivessnake/edit_notes.py:
```diff
@@ -7,64 +7,67 @@
-                    for subnote in note.get("subnotes"):
-                        content = subnote["content"]
+                if note['type'] == args.note_type:
+                    for subnote in note.get('subnotes'):
+                        content = subnote['content']
```
```suggestion
                    for subnote in note.get('subnotes', []):
                        content = subnote['content']
```
As written, line 31 will error if there are no subnotes because the `get` method returns `None` by default if the key is missing. Adding a second argument returns that value instead, so I've added an empty list.

helrond commented on archivessnake/edit_notes.py:
```diff
@@ -7,64 +7,67 @@
-                            updated = True
-            if updated:
-                save_record(record.uri, aojson)
+                            handle_matching_notes(args, notes, idx, record, content, subnote, client, resource_id, writer, updated)
```
Looking at this, you are passing a bunch of arguments into this function just so you can pass them from there into another function. That indicates to me that there's an opportunity for rethinking the logic here. Either my suggestion for breaking this out into a separate method was bad (possible) or `handle_matching_notes` just needs to do stuff with the notes and return the updated data, which you can then write to a CSV and save in AS when necessary.

helrond commented on archivessnake/edit_notes.py:
```diff
@@ -7,64 +7,67 @@
+    log_to_spreadsheet(record, writer, resource_id, client)
+    updated = True
+    if updated:
+        save_record(client, record['uri'], record)
```
As per my comment above, these lines do not semantically seem to me to be a part of `handle_matching_notes`.

helrond commented on archivessnake/edit_notes.py:
```diff
@@ -7,64 +7,67 @@
-            top_container.type.capitalize(), top_container.indicator)
-        writer.writerow([container, archival_object.uri])
+    if record['instances']:
+        for instance in record['instances']:
```
```suggestion
    for instance in record.get('instances', []):
```
As above, you can make this a bit more concise and avoid KeyErrors by using `get` with a second argument.

helrond commented on archivessnake/edit_notes.py:
```diff
@@ -7,64 +7,67 @@
-        container = "{} {}".format(
-            top_container.type.capitalize(), top_container.indicator)
-        writer.writerow([container, archival_object.uri])
+    if record['instances']:
```
I would consider making this container bit its own method because semantically it's different than logging to a spreadsheet.

helrond commented on archivessnake/edit_notes.py:
```diff
@@ -7,64 +7,67 @@
+            top_container['type'].capitalize(), top_container['indicator'])
+    else: 
+        container = "None"
+    writer.writerow([resource_id['title'], resource_id['id_0'], record['uri'], record['ref_id'], record['title'], container])
```
```suggestion
    writer.writerow([resource_id['title'], resource_id['id_0'], record['uri'], record['ref_id'], record.get('title'), container])
```
I am pretty sure that it's possible for AS objects to not have a title if they have a date, so let's accommodate that edge case.

helrond commented on archivessnake/edit_notes.py:
```diff
@@ -7,64 +7,67 @@
+    if updated:
+        save_record(client, record['uri'], record)
 
 def contains_match(content, search_string):
     """Returns True if user-provided note input matches the corresponding note within a given ratio (CONFIDENCE_RATIO)."""
     ratio = fuzz.token_sort_ratio(content.lower(), search_string.lower())
     return True if ratio > CONFIDENCE_RATIO else False
```
I think we can use the ASnake method `text_in_note` here, right? https://github.com/archivesspace-labs/ArchivesSnake/blob/master/asnake/utils/__init__.py#L106 And if I'm reading it right it might help us eliminate some of the above conditionals as well?

helrond reviewed (commented):
Added a bunch of comments here - I think the work you've done has clarified for me where the overall structure is a bit confusing (and will be hard to maintain). Basically I think there are a bunch of nested functions which should be pulled out more as siblings. It may be helpful to write out the steps of the process to get a sense of how to put it all together.

kemartin2015 commented:
Hi @helrond, I think I've addressed the issues in your previous round of comments, but I have some more questions now:

1. Now that I’ve untangled some of these methods, I’m not sure how to get the data back to `main.` Does it make sense to move where `resource` is defined and `walk_tree `to the main method to alleviate some of the passing issues? I’m also not confident in the logic in `main`.
2. I’m still passing quite a few arguments into `handle_matching_notes` even after moving `save_record` out. Is that okay or should these lines be moved back into `process_tree`? 
3. I’m getting an invalid syntax error when I try to pass `args.resource_id` into `process_tree`. Do I need to rename the `args.` arguments to prevent that error?
4. I added in the ASnake text_in_note method, and to me, it looks like a simpler way to do what we were already doing, but I’m not sure if I’m using it correctly?

kemartin2015 commented:
Hi @helrond, I'm getting a couple errors when I run the script.

In line 60, i get this error:

`TypeError: argument 1 must have a "write" method`

In line 26, I was getting the following error. Is the issue with the `content` argument? The script runs if I change `content` to `note` in line 26, but then it modifies/deletes every access note even with a high confidence ratio.
 
````
Traceback (most recent call last): 
  File "/Users/k.martin/Documents/GitHub/scripts/archivessnake/edit_notes.py", line 83, in <module> 
    main(args.note_type, args.action, args.resource_id, args.search_string, args.level, args.replace_string) 
  File "/Users/k.martin/Documents/GitHub/scripts/archivessnake/edit_notes.py", line 65, in main 
    processed_record, updated = process_record(record, level, note_type, action, search_string, replace_string) 
  File "/Users/k.martin/Documents/GitHub/scripts/archivessnake/edit_notes.py", line 26, in process_record 
    if text_in_note(content, search_string, CONFIDENCE_RATIO): 
  File "/opt/homebrew/lib/python3.9/site-packages/asnake/utils/__init__.py", line 117, in text_in_note 
    note = resolve_to_json(note, client) 
  File "/opt/homebrew/lib/python3.9/site-packages/asnake/utils/__init__.py", line 49, in resolve_to_json 
    json = client.get(thingit).json() 

AttributeError: 'int' object has no attribute 'get'
````

helrond commented:
OK, there was an error with the way the CSV Writer was being created; I've fixed that now and that resolves the issue in line 60.

The other issue is slightly more complex. It looks like the `text_in_note` utility is expecting that you will pass a note object, a query string, an ASpace client and optionally a confidence ratio but instead we are passing subnote content and a query string. HOWEVER we can't just move that conditional further up and update the arguments because you're replacing content in a specific subnote. So it might actually make sense here to go back to the way you had it originally, even though it's somewhat duplicative of what ASnake can do.

kemartin2015 commented:
Ok, sounds good! Thanks for explaining the `text_in_note` issue. That makes a lot of sense.

kemartin2015 commented:
Hi @helrond, I reverted `text_in_note` back to `contains_match,` but I don't think your change to csv writer was committed because I'm still getting that error.

helrond commented:
Ah, sorry I forgot to push. Changes are there now.

kemartin2015 commented:
I tested the script on FA739A and FA739B in AS dev, and it looks good to me. I didn't have issues modifying or deleting access notes. The output CVSs matched the original output of the notes_csv,py script, and I didn't see any unexpected changes in other notes or other areas of the finding aid. Is there anything else I should be testing for?

helrond commented:
I don't think so? We mostly just moved things around so they're a little more atomic. At some point in the future we could also write tests for these scripts (particularly the ones which are writing data back to AS) so we have a better safety net.

kemartin2015 commented:
Ok, sounds good! Should I go ahead and merge these changes?

helrond commented:
Sure, go ahead and merge!
